### PR TITLE
fixed hf convert and now it's working with tinyllama

### PR DESCRIPTION
**Pull request by SkullMag**
Fixed huggingface convert

### CODE REVIEW

**Comment by SkullMag** [comment]
I've compared the outputs with the transformers code:
```python
from transformers import LlamaForCausalLM, LlamaTokenizer

model = LlamaForCausalLM.from_pretrained("tinyllama_503b", local_files_only=True)
tokenizer = LlamaTokenizer.from_pretrained("tinyllama_503b/tokenizer.model")
prompt = "Where do you live?"
inputs = tokenizer(prompt, return_tensors="pt")
generate_ids = model.generate(inputs.input_ids, max_length=30)
print(tokenizer.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0])
```

Here is the transformers output:
```
The 2018-19 season is the 10th season for the club and the
```

And here is the output from tinygrad:
```
python llama.py --gen=tiny --model tinyllama_503b/model.safetensors --size 1B --temperature 0 --count 30 --prompt "Where do you live?"

The 2018-19 season is the 10th season for the club and the 10th season in the
```

**Comment by SkullMag** [comment]
I used the config from https://github.com/tinygrad/tinygrad/issues/2228

File: examples/llama.py [hunk 1]
```diff
@@ -218,6 +218,12 @@ def __call__(self, tokens:Tensor, start_pos:Variable, temperature:float=0.0):
       "args": {"dim": 8192, "n_layers": 48, "n_heads": 64, "n_kv_heads": 8, "multiple_of": 256, "ffn_dim_multiplier": 1.0, "norm_eps": 1e-5, "rope_theta": 1000000, "vocab_size": 32000},
       "files": 4,
     },
+  },
+  "tiny": {
+    "1B": {
+      "args": {"dim": 2048, "multiple_of": 256, "n_heads": 32, "n_kv_heads": 4, "n_layers": 22, "norm_eps": 1e-05, "vocab_size": 32000},
```
Comment (chenyuxyz) [inline examples/llama.py:224]: nit: can you order this to have the same order as the ones above?

File: examples/llama.py [hunk 2]
```diff
@@ -252,7 +261,21 @@ def convert_from_huggingface(weights, model):
...
-  return {keymap[k]: v for k,v in weights.items() if ".rotary_emb." not in k}
+  sd = {}
+  for k, v in weights.items():
+    if ".rotary_emb." in k:
+      continue
+    v = v.to(Device.DEFAULT)
+    to_name = keymap[k]
+    if "model.layers" in k:
+      if "q_proj" in k:
+        sd[to_name] = permute(v, 32, 2048, 2048)
```
Comment (chenyuxyz) [inline examples/llama.py:272]: use the params instead of hard coding the numbers

File: examples/llama.py (hunk 1 above)
Comment (SkullMag) [reply examples/llama.py]: done

File: examples/llama.py (hunk 2 above)
Comment (SkullMag) [reply examples/llama.py]: done

File: examples/llama.py [hunk 3]
```diff
@@ -242,7 +248,10 @@ def load(fn:str):
   else:
     return torch_load(fn)
 
-def convert_from_huggingface(weights, model):
+def convert_from_huggingface(weights, model: Transformer, n_heads: int, n_kv_heads: int):
+  def permute(v: Tensor, n_heads: int, shape):
```
Comment (chenyuxyz) [inline examples/llama.py:252]: do you need `shape` if it's always `v.shape` when called?

File: examples/llama.py [hunk 4]
```diff
@@ -489,7 +513,7 @@ def main(string: str):
 
   # *** prompt engineers stop here ****
 
-  LLAMA_SUFFIX = {"1": "", "2": "-2", "code": "-code"}[args.gen]
+  LLAMA_SUFFIX = {"1": "", "2": "-2", "code": "-code", "tiny": "tiny"}[args.gen]
```
Comment (chenyuxyz) [inline examples/llama.py:516]: "-tiny"

Comment (SkullMag) [reply examples/llama.py]: fixed

**Comment by geohot** [comment]
Can it auto fetch the weights if they aren't there?

**Comment by SkullMag** [comment]
@geohot @chenyuxyz I've added default size evaluation based on model gen and added code for dynamically updating help for easier development in the future. I suggest we add model fetch in separate PR, what do you think?

**Comment by geohot** [comment]
Tested and merged, george@tinygrad.org to claim the bounty, either USDC or PayPal.

**Pull request merged**